packages/xpetra/src/CrsMatrix/Xpetra_TpetraCrsMatrix_def.hpp

```diff
@@ -0,0 +1,473 @@
+// @HEADER
+//
+// ***********************************************************************
+//
+//             Xpetra: A linear algebra interface package
+//                  Copyright 2012 Sandia Corporation
+//
+// Under the terms of Contract DE-AC04-94AL85000 with Sandia Corporation,
+// the U.S. Government retains certain rights in this software.
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are
+// met:
+//
+// 1. Redistributions of source code must retain the above copyright
+// notice, this list of conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright
+// notice, this list of conditions and the following disclaimer in the
+// documentation and/or other materials provided with the distribution.
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY SANDIA CORPORATION "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL SANDIA CORPORATION OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef XPETRA_TPETRACRSMATRIX_DEF_HPP
+#define XPETRA_TPETRACRSMATRIX_DEF_HPP
+
+#include "Xpetra_TpetraCrsMatrix_decl.hpp"
```

Comment:
The `_def.hpp` headers should not need to include the `_decl.hpp` headers, for the following reasons:

  1. Automatically generated `.hpp` header files will do the right thing, whether or not ETI is enabled.  If ETI is ON, they include only the corresponding decl file.  If ETI is OFF, they include first the decl file, then the def file.
  2. This means that if ETI is OFF, users will double-include the def file.  This has a nonzero cost (the C preprocessor generally must reread the file).

Removing the include means that your automatically generated .cpp ETI files must include both the decl and the def header files.  Right now, they just include the def header.

Reply:
Huh -- doesn't the *_def.hpp need to know about the class declaration?

Reply:
@jhux2 Nothing ever includes the `${FOO}_def.hpp` file by itself, without `${FOO}_decl.hpp` first having been included.  Here are the possible scenarios:

1. With ETI OFF, user includes `${FOO}.hpp`, which includes first `${FOO}_decl.hpp`, then `${FOO}_def.hpp`.
2. With ETI ON, user includes `${FOO}.hpp`, which includes just `${FOO}_decl.hpp`.
3. User includes just `${FOO}_decl.hpp` for the class declaration.  At some later point, user may include `${FOO}.hpp`.  See above.
4. With ETI ON, the ETI `.cpp` file includes first `${FOO}_decl.hpp`, then `${FOO}_def.hpp`.

Reply:
@mhoemmen You are assuming CMake generated `Foo.hpp`, and that the source is partitioned into `Foo_decl.hpp` and `Foo_def.hpp`. Seems your comment is really rooted in how CMake enables/disables ETI.

How does MueLu/Xpetra handle this?  When I grepped the source, I saw lots of "Foo.hpp", which seems to indicate those files are not generated by CMake.

Reply:
@jjellio Good point -- I was assuming that `${FOO}.hpp` includes just the decl if ETI is ON, and includes first the decl and then the def if ETI is OFF.  As long as MueLu's `${FOO}.hpp` header files look like that, what I wrote above should still apply.